FD 1.17.0: Documentation other than the release notes (#135)
mulesoft/docs-design-center: modules/ROOT/pages/fd-iterating.adoc

Counter Variable Name:: Name of the variable that stores the index of the element that is being processed. For example, if a collection is `['a','b','c']`, the index of `a` is 0, of `b` is 1, and of `c` is 2. When the For-Each component is processing the second element, the value of the variable is 1.

== Configuring a Parallel For-Each Component

When you configure this component, you must specify values in these fields:

Collection:: Specifies the expression that defines the collection of parts to be processed in parallel. The default is `payload`.

Timeout:: Specifies the maximum time in milliseconds that the entire scope of a Parallel For-Each component can take to complete processing all of the elements in a collection.

Max Concurrency:: Specifies the maximum number of parallel processes that can be run. How many parallel processes are in fact started depends on how Mule Runtime allocates resources where it is running:

* If you set the value to 40, but Mule Runtime has allocated only enough resources to run up to 4 processes in parallel, only up to 4 processes can run in parallel.

> * If you set the value to 2, but Mule Runtime has allocated enough resources to run up to 4 processes in parallel, only up to 2 processes can run in parallel

Reviewer: Mule or Mule runtime engine, not Mule Runtime.

= Process Elements in a Collection
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

:imagesdir: ../assets/images

> You can process elements in a collection by using the For-Each component or the Parallel For-Each component. Each component sets a scope within which you can add other types of component, as well as connectors and modules, for processing each element.

Author: For-Each and Parallel For-Each are types of component. They are documented currently only in the documentation about the MuleSoft core components.

> You can process elements in a collection by using the For-Each component or the Parallel For-Each component. Each component sets a scope within which you can add other types of component, as well as connectors and modules, for processing each element.

Reviewer 2: If the definition of component, module and connector is known or is written in another doc, then the way we're making reference of them seems fine, if not, it's a bit confusing. In Mule 4 there's no distinction between stuff, everything is an extension. In Mule 3 there were connectors and such.

.The For-Each component and the Parallel For-Each component both set scopes for processing the elements within a collection.
image::fd-iterators-example.png[Two simple examples: one of the use of For-Each, the other of the use of Parallel For-Each]

<1> The For-Each component in this example inserts each element, one at a time, into a database. Each element is presumably a string that can be used in a SQL INSERT statement.
<2> The Parallel For-Each component in this example sends each element to a Set Payload component that modifies the element with an expression and adds it to the flow's payload.

> Both types of component can process any collection. A collection can be any supported content type, such as application/json, application/java, or application/xml.

Reviewer 2: I would remove the following:
```
 A collection can be any supported content type, such as application/json, application/java, or application/xml.
```
Because a collection doesn't have an application type per se.

> Both types of component can process any collection, including lists and arrays. A collection can be any supported content type, such as application/json, application/java, or application/xml. For an XML or JSON collection, specify a DataWeave expression in the *Collection* field to split the collection into its constituent elements. The For-Each or Parallel For-Each scope can split a Java collection into its elements without depending on a DataWeave expression to explain how to split it.

Author: I've removed the last two sentences of the paragraph. Larger problem: How thoroughly  do I have to explain what a collection is?

> Both types of component can process any collection, including lists and arrays. A collection can be any supported content type, such as application/json, application/java, or application/xml. For an XML or JSON collection, specify a DataWeave expression in the *Collection* field to split the collection into its constituent elements. The For-Each or Parallel For-Each scope can split a Java collection into its elements without depending on a DataWeave expression to explain how to split it.

Reviewer 2: Yes, it should be enough. The for each card needs the user to understand that they will be iterating over a collection of items and performing some actions on their items.

== How the For-Each Component Processes Data and Continues a Flow

The For-Each component is for processing each element in a collection serially and then proceeding in the flow with the original payload that was passed to the For-Each component:

.This diagram represents a flow in which a For-Each component is used. The rounded rectangles are cards.
image::fd-foreach.png[Diagram depicting the actions taken by the For-Each component]

<1> A card passes the payload of a flow to the For-Each component. The payload consists of a collection of elements.

> <2> The For-Each component divides the collection into separate elements. It then sends each element serially to the first component, connector, or module that is in its scope. Each element is treated as a payload within the scope. So, if a connector within the scope is configured to act on `payload` or uses a script to process `payload`, the payload is not the payload that the For-Each component received.

Reviewer 2: ```suggestion
<2> The For-Each component divides the collection into separate elements. It then sends each element serially to the first component, connector, or module that is in its scope. Each element is treated as a payload within the scope. So, if a connector within the scope is configured to act on `payload` or uses a script to process `payload`, the payload is not the payload that the For-Each component received but the element it is processing.
```
I would clarify that payload is the element it is processing.

<3> After the final element is processed, the For-Each component passes the payload that it received from the previous card to the next card in the flow.
<4> The next card in the flow receives the same payload that the For-Each component received, and the flow continues.

== How the Parallel For-Each Component Processes Data and Continues a Flow

The Parallel For-Each component is for processing each element in a collection and then passing the results to the next card in the flow:

.This diagram represents a flow in which a Parallel For-Each component is used. The rounded rectangles are cards.
image::fd-parallel-foreach.png[Diagram depicting the actions taken by the Parallel For-Each component]

<1> A card passes the payload of a flow to the Parallel For-Each component. The payload consists of a collection of elements.

> <2> The Parallel For-Each component divides the collection into separate elements. It then sends elements simultaneously to the first component, connector, or module that is in its scope. You can specify the maximum number of elements to process at a time. Each element is treated as a payload within the scope. So, if a connector within the scope is configured to act on `payload` or uses a script to process `payload`, the payload is not the payload that the Parallel For-Each component received.

Reviewer 2: ```suggestion
<2> The Parallel For-Each component divides the collection into separate elements. It then sends elements simultaneously to the first component, connector, or module that is in its scope. You can specify the maximum number of elements to process at a time. Each element is treated as a payload within the scope. So, if a connector within the scope is configured to act on `payload` or uses a script to process `payload`, the payload is not the payload that the Parallel For-Each component received but the element of the collection it is processing.
```
I would clarify what payload is here.

[IMPORTANT]
====
Within the scope of the Parallel For-Each component, the order in which each element is processed is not guaranteed to be the order in which each element appears in a collection. Therefore, if any of the connectors require elements in a particular order (for example, if you want to update database tables in a particular order with the Database connector), you should use the For-Each component instead.

Moreover, you should be aware whether any APIs or services that you want to connect to within the scope of a Parallel For-Each component have any limits in the number of concurrent connections that they accept. If they do, you must specify a maximum number of concurrent processes for the Parallel For-Each component.
====

// == Transactional Processing

> // The For-Each component does nothing differently when processing a transaction.

Reviewer 3: differently from what?

> However, the Parallel For-Each component does not process the elements of a collection simultaneously when processing a transaction. Instead, it processes them serially, just as the For-Each component does. The output from this component is still the aggregated results from the processing of the elements in a collection.

Reviewer 2: What do we mean when we talk about a transaction? I'm not familiar with the wording.
Can we provide an example here?

Author: I've removed this section for now. I'm also not yet familiar with what a "transaction" means in this context.

== Batching

With the For-Each component, you can split a collection into batches to enable quicker processing. Each batch is treated as a separate Mule message. For example, if a collection has 200 elements and you set Batch Size to 50, the For-Each component iteratively processes 4 batches of 50 elements, each as a separate Mule message.

Batching is not supported with the Parallel For-Each component.

== Variables

> Every iteration within the For Each component can make use of the variables that were used in the previous iteration. New variables or modifications to existing variables that take place during the processing of one element are visible during the processing of another element. These changes to variables continue to be available outside the For Each scope.

Reviewer 3: ```suggestion
Every iteration within the For-Each component can use the variables that were used in the previous iteration. New variables or modifications to existing variables that take place during the processing of one element are visible during the processing of another element. These changes to variables continue to be available outside the For-Each scope.
```

For example, suppose that you use a Set Variable component just before the scope of a For-Each component. Within the Set Variable component, you set the variable `NumOfElements` to 0. You then use a Set Variable component in the For-Each scope, using the same variable, that specifies a DataWeave script that increments the value by 1 every time it is called. Every time that the For-Each component sends another element of a collection to be processed, the value is incremented and saved in `NumOfElements`. When all of the elements have been processed, the value of `NumOfElements` is the number of elements that were in the collection. The value can be used outside of the For-Each scope by subsequent cards in the flow.

> For the Parallel For Each component, the value of a variable that enters the scope is the same at the start of the processing of each variable. New variables or modifications of already existing variables during the processing of one element are not visible during the processing of another element. Moreover, no changes to variables with the scope of a Parellel For-Each component are available to the subsequent cards in the flow.

Reviewer 3: ```suggestion
For the Parallel For-Each component, the value of a variable that enters the scope is the same at the start of the processing of each variable. New variables or modifications of already existing variables during the processing of one element are not visible during the processing of another element. Moreover, no changes to variables with the scope of a Parallel For-Each component are available to the subsequent cards in the flow.
```

== Error Handling

Both the For-Each component and Parallel For-Each component stop processing a collection if an element causes any component, connector, or module in a scope to throw an exception. In addition, the Parallel For-Each component stops processing a collection if one of the parallel processes times out while waiting for an open connection to a system. You can set the maximum length of time that a parallel process can wait.

When the processing of a collection stops because of an exception, Flow Designer displays a red circle on the card that threw the exception. When the processing of a collection stops because of a timeout, Flow Designer displays a red circle on the Parallel For-Each component's card. You can open the *Logs* pane to read the error message.

== Configuring a For-Each Component

When you configure this component, you must specify values in these fields:

Collection:: An expression that returns a Java collection, object array, map, or DOM nodes. The default is `payload`.

> Batch Size:: Partitions the collection into sub-collections of the specified number of elements. For example, if a collection has 200 elements and you set the batch size to 50 elements, the component processes 4 batches of 50 elements each.

Reviewer 3: ```suggestion
Batch Size:: Partitions the collection into subcollections of the specified number of elements. For example, if a collection has 200 elements and you set the batch size to 50 elements, the component processes 4 batches of 50 elements each.
```

> Root Message Variable Name:: Name of the property that stores the entire message -- payload, variables, and attributes -- that was passed to a For-Each component from the previous card in a flow. 

Reviewer 3: GLOBALLY replace dashes or em-dashes formed using spaces and hyphens with punctuation that has the same or similar meaning. In this case, I would use parentheses:
```suggestion
Root Message Variable Name:: Name of the property that stores the entire message (payload, variables, and attributes) that was passed to a For-Each component from the previous card in a flow. 
```

> Root Message Variable Name:: Name of the property that stores the entire message -- payload, variables, and attributes -- that was passed to a For-Each component from the previous card in a flow. 

Reviewer 2: ```suggestion
Root Message Variable Name:: Name of the variable that stores the entire message -- payload, variables, and attributes -- that was passed to a For-Each component from the previous card in a flow. 
```
I would not use property here as it's a variable name. Property might be confused with an element of an object.

Reviewer 2: The same applies to the counter variable name.

> Root Message Variable Name:: Name of the property that stores the parent message. The parent is the complete, non-split message. // What is the root message variable name for? What does it store "the parent message" for? Is "parent message" the full collection?

Reviewer 2: The root message variable name is the name of the variable that will be set to be used inside the for each scope to access the original message. The original message is the whole message that a flow processes, that includes the payload, variables, etc.

> Counter Variable Name:: Name of the property that stores the index of the element that is being processed. For example, if a collection is `['a','b','c']`, the index of `a` is 0, of `b` 1, and of `c` 2. When the For-Each component is processing the second element, the value of the variable is 1.

Reviewer 3: For internationalization:
```suggestion
Counter Variable Name:: Name of the property that stores the index of the element that is being processed. For example, if a collection is `['a','b','c']`, the index of `a` is 0, of `b` is 1, and of `c` is 2. When the For-Each component is processing the second element, the value of the variable is 1.
```

> Counter Variable Name:: Name of the property that stores the number of messages over which it iterates. // What is "it"?

Reviewer 2: The counter variable name is the name of the variable that will be set to be used inside the for each scope to identify the index of the element of the collection being processed. That is, if the for each receives as collection like [1, 2, 3, 4, 5] and it's processing the second element (2), the variable's value will be 1 (because it has already processed the element 0 and it's now processing the element 1).

== Configuring a Parallel For-Each Component

When you configure this component, you must specify values in these fields:

Collection:: Specifies the expression that defines the collection of parts to be processed in parallel. The default is `payload`.

> Timeout:: Specifies the maximum time in milliseconds that a parallel process can wait for an open connection to a system reached via a connector.

Reviewer 2: This is the timeout of the execution of the whole scope in the parallel foreach. It can timeout at any time, in any card. That is, the timeout there is the maximum time the for each can take to execute the scope.

> Max Concurrency:: Specifies the maximum number of parallel processes that can be run. How many parallel processes are in fact started depends on how Mule Runtime allocates resources where it is running.

Reviewer 3: ```suggestion
Max Concurrency:: Specifies the maximum number of parallel processes that can be run. How many parallel processes are in fact started depends on how Mule Runtime allocates resources where it is running:
```

> * If you set the value to 40, but Mule Runtime has allocated only enough resources to run up to 4 processes in parallel, only up to 4 processes can run in parallel.

Reviewer 3: ```suggestion
* If you set the value to 40, but Mule has allocated only enough resources to run up to 4 processes in parallel, only up to 4 processes can run in parallel.
```

> * If you set the value to 2, but Mule runtime has allocated enough resources to run up to 4 processes in parallel, only up to 2 processes can run in parallel

Reviewer 3: ```suggestion
* If you set the value to 2, but Mule has allocated enough resources to run up to 4 processes in parallel, only up to 2 processes can run in parallel.
```
